Package.swift

```diff
@@ -108,7 +108,21 @@ var targets: [PackageDescription.Target] = [
             "NIOSOCKS",
             .product(name: "NIOCore", package: "swift-nio"),
             .product(name: "NIOEmbedded", package: "swift-nio"),
-        ])
+        ]),
+    .target(
+        name: "NIONFS3",
+        dependencies: [
+            .product(name: "NIO", package: "swift-nio"),
```

Comment:
Why depend on umbrella NIO? IIRC we want to deprecate importing the umbrella module?

Reply:
@fabianfett Honestly, I think that's the right thing to do because given how SwiftPM works it's just too easy to forget a dependency and not be told about it. It also feels wrong to burden normal users with even more NIO imports. But I'll follow the remaining code here and add the NIO{Embedded,Core,Posix,...} imports.

Reply:
@weissi FYI: Swift 5.8 has a new flag to tell you if you forget a dependency `--explicit-target-dependency-import-check error`. We have enabled it for the nightly CI for some time now and [we added Swift 5.8. CI today with it enabled. ](https://github.com/apple/swift-nio-extras/blob/e4efedbf47554258e85e6c3949fc956fa6187c76/docker/docker-compose.2204.58.yaml#L14) IIRC the plan is to enable this in Swift 6 by default.

Reply:
Sure, I'm aware of that but I don't think it's reasonable to force users to list NIOCore, NIOPosix, NIOEmbedded, ... etc when previously just listing `NIO` was correct?

Reply:
I strongly disagree with respect to what `import NIO` means today. The set of reexported libraries (NIOCore, NIOPosix, NIOEmbedded) only have historic reasons because they once were all in the original `NIO` module and were not chosen on purpose. Therefore it is only very rarely what you actually want as `NIOEmbedded` is usually only used in tests and not in the actually implementation. The current PR is actually a good example for this. The `NIO` dependency of `NIONFS3` was not replaced with `NIOCore`, `NIOPosix` and `NIOEmbedded` but just with `NIOCore`. The test target has replaced `NIO` with `NIOCore` and `NIOEmbedded`.

Umbrella modules are not really a common thing in Swift. The only way to find out that it is an umbrella module and which libraries it covers is to either look in the documentation or in the Package.swift and implementation. Tooling isn't great either.

I agree that it is anyone to have lots of imports. I don't have a good answer today though. Maybe we should design one or more umbrella modules which make this easier but I can't come up with a good set which include more than 2 libraries. Maybe `NIOPosix`, `NIOHTTP1`, `NIOWebSocket` etc. should just reexport `NIOCore`.

Reply:
Also, what if `NIOCore` turns out to not be small enough? It might be split even further but `NIOCore` has to continue to work, much like `NIO` today.
Maybe the SwiftPM feature which allows you to export a product that then exports multiple modules solves part or this issue.

This definitely needs exploration on how we can square usability and technical needs (like not necessarily bringing in NIOPosix)

Reply:
Just for the record: happy to change in this PR but I think this is the wrong direction and just messes with users to now have to decide if they want NIO, NIOCore, NIOPosix or maybe something else. They have no idea what NIOPosix or NIOCore even is but they do know `import NIO` because it’s everywhere :)

Reply:
The problem with this is simply one of compatibility. If you bring NIOPosix with you then you force everyone to compile it, a source of substantial pain for users who don't need or want it. It bloats binary size, increases compile times, and generally makes people's lives harder. We shouldn't make people pay for what they aren't using.

The NIO umbrella module exists for end-users who want the umbrella. For people who are aware of the differences, we should enable the advanced use-case.

Reply:
As I said, happy to changed _here_ because it's a NIO package and it's fine to mess with NIO devs (because there are legitimate special case users that don't want posix). But in general, the `NIO` umbrella should be fully supported and the default way to pull in NIO IMHO.